pandas/core/arrays/base.py

```diff
@@ -216,6 +218,88 @@ def isna(self):
         """
         raise AbstractMethodError(self)
 
+    def tolist(self):
+        # type: () -> list
+        """Convert the array to a list of scalars."""
+        return list(self)
+
+    def fillna(self, value=None, method=None, limit=None):
+        """ Fill NA/NaN values using the specified method.
+
+        Parameters
+        ----------
+        method : {'backfill', 'bfill', 'pad', 'ffill', None}, default None
+            Method to use for filling holes in reindexed Series
+            pad / ffill: propagate last valid observation forward to next valid
+            backfill / bfill: use NEXT valid observation to fill gap
+        value : scalar, array-like
+            If a scalar value is passed it is used to fill all missing values.
+            Alternatively, an array-like 'value' can be given. It's expected
+            that the array-like have the same length as 'self'.
+        limit : int, default None
+            (Not implemented yet for ExtensionArray!)
+            If method is specified, this is the maximum number of consecutive
+            NaN values to forward/backward fill. In other words, if there is
+            a gap with more than this number of consecutive NaNs, it will only
+            be partially filled. If method is not specified, this is the
+            maximum number of entries along the entire axis where NaNs will be
+            filled.
+
+        Returns
+        -------
+        filled : ExtensionArray with NA/NaN filled
+        """
+        from pandas.api.types import is_scalar
+        from pandas.util._validators import validate_fillna_kwargs
+
+        value, method = validate_fillna_kwargs(value, method)
+
+        if not is_scalar(value):
+            if len(value) != len(self):
+                raise ValueError("Length of 'value' does not match. Got ({}) "
+                                 " expected {}".format(len(value), len(self)))
+        else:
+            value = itertools.cycle([value])
+
+        if limit is not None:
+            msg = ("Specifying 'limit' for 'fillna' has not been implemented "
+                   "yet for {} typed data".format(self.dtype))
+            raise NotImplementedError(msg)
+
+        mask = self.isna()
+
+        if mask.any():
+            # ffill / bfill
+            if method is not None:
+                if method == 'backfill':
+                    data = reversed(self)
+                    mask = reversed(mask)
+                    last_valid = self[len(self) - 1]
+                else:
+                    last_valid = self[0]
+                    data = self
+
+                new_values = []
+
+                for is_na, val in zip(mask, data):
```

Comment:
huh? what is all this.

Reply:
As you can clearly see from the diff, it is an generic implementation of `fillna`, and the above part for the `method` kwarg (forward or backward fill). 
If you don't like the implementation, it can be more helpful to give a more specific comment than that.

@TomAugspurger do you want this for cyberpandas? Otherwise, we could also start for now with only supporting filling with a `value` ? (that's what we supported for now in geopandas)

Reply:
On the other hand, we could maybe do this differently (that *might* be more efficient, without looping through all data): create an array of indices (arange), mask them with the missing values, then fill those indices with the existing fill-method machinery, and then use that to index into the original data and to fill the missing values.

Reply:
I don't really care about ffill / bfill, but Categorical supports it so best to be consistent.

Here's one alternative based astyping to object and setting.

```diff
+        from pandas.core.missing import pad_1d, backfill_1d
 
         value, method = validate_fillna_kwargs(value, method)
 
+        mask = self.isna()
+
         if not is_scalar(value):
             if len(value) != len(self):
                 raise ValueError("Length of 'value' does not match. Got ({}) "
                                  " expected {}".format(len(value), len(self)))
-        else:
-            value = itertools.cycle([value])
+            value = value[mask]
 
         if limit is not None:
             msg = ("Specifying 'limit' for 'fillna' has not been implemented "
                    "yet for {} typed data".format(self.dtype))
             raise NotImplementedError(msg)
 
-        mask = self.isna()
-
         if mask.any():
             # ffill / bfill
-            if method is not None:
-                if method == 'backfill':
-                    data = reversed(self)
-                    mask = reversed(mask)
-                    last_valid = self[len(self) - 1]
-                else:
-                    last_valid = self[0]
-                    data = self
-
-                new_values = []
-
-                for is_na, val in zip(mask, data):
-                    if is_na:
-                        new_values.append(last_valid)
-                    else:
-                        new_values.append(val)
-                        last_valid = val
-
-                if method in {'bfill', 'backfill'}:
-                    new_values = list(reversed(new_values))
+            if method == 'pad':
+                values = self.astype(object)
+                new_values = pad_1d(values, mask=mask)
+            elif method == 'backfill':
+                values = self.astype(object)
+                new_values = backfill_1d(values, mask=mask)
             else:
                 # fill with value
-                new_values = [
-                    val if is_na else original
-                    for is_na, original, val in zip(mask, self, value)
-                ]
+                new_values = self.copy()
+                new_values[mask] = value
         else:
             new_values = self
         return type(self)(new_values)
```

Timings:

astype

```
1000, value, 0.01
1000, ffill, 0.00
1000, bfill, 0.00
1000000, value, 4.61
1000000, ffill, 2.34
1000000, bfill, 2.36
```

Loopy

```
1000, value, 0.00
1000, ffill, 0.00
1000, bfill, 0.00
1000000, value, 1.28
1000000, ffill, 1.37
1000000, bfill, 1.32
```

So the loopy version (current PR) is ~2-4x faster. Lots of variability there, esp depending on the speed of EA.__setitem__, but it seems like a good default.

Reply:
What code did you use for the timings here?

Reply:
```python
from timeit import default_timer as tic
import pandas as pd
import numpy as np
from pandas.tests.extension.decimal.array import DecimalArray
from decimal import Decimal

Ns = [1000, 1_000_000]

for N in Ns:

    arr = np.random.uniform(size=(N,))
    arr[arr > .8] = np.nan
    arr = DecimalArray([Decimal(x) for x in arr])

    start = tic()
    arr.fillna(value=Decimal('1.0'))
    stop = tic()

    print(f'{N}, value, {stop - start:0.2f}')

    start = tic()
    arr.fillna(method='ffill')
    stop = tic()

    print(f'{N}, ffill, {stop - start:0.2f}')

    start = tic()
    arr.fillna(method='bfill')
    stop = tic()

    print(f'{N}, bfill, {stop - start:0.2f}')
```

I can re-run it / profile things a bit to see why the array[object] version was so much slower.

Reply:
With a quick dirty check:

```
In [1]: from geopandas.array import from_shapely
   ...: import shapely.geometry
   ...: a = np.array([shapely.geometry.Point(i, i) for i in np.random.randn(10000)], dtype=object)
   ...: a[np.random.randint(0, 10000, 10)] = None
   ...: ga = from_shapely(a)
   ...: s = pd.Series(ga)

In [2]: s.head()
Out[2]: 
0    POINT (-0.9684702137574125 -0.9684702137574125)
1        POINT (0.419808066032094 0.419808066032094)
2          POINT (0.57207855258413 0.57207855258413)
3      POINT (0.7773915414611537 0.7773915414611537)
4        POINT (2.073261603753559 2.073261603753559)
dtype: geometry

In [3]: %timeit s.fillna(method='pad')
115 ms ± 730 µs per loop (mean ± std. dev. of 7 runs, 10 loops each)

In [4]: %timeit s.fillna(method='pad_new')
217 µs ± 3.27 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)

In [5]: s.fillna(method='pad').equals(s.fillna(method='pad_new'))
Out[5]: True
```

And for this 'new' method I did :

```
+                if method == 'pad_new':
+                    idx = np.arange(len(self), dtype=float)
+                    idx[mask] = np.nan
+                    idx = pad_1d(idx)
+                    idx = idx.astype('int64')
+                    new_values = self.take(idx)
+                else: ... existing code in PR
```

Reply:
OK, didn't see your post before posting mine. So doing the same with the decimal array `arr` (with 10000 values) doesn't show a speed-up, rather a slowdown:

```
In [18]: %timeit arr.fillna(method='pad')
17.3 ms ± 772 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)

In [19]: %timeit arr.fillna(method='new')
28.5 ms ± 116 µs per loop (mean ± std. dev. of 7 runs, 10 loops each)
```

I suppose this is because DecimalArray uses an object array under the hood anyhow? So something like `take` will not be that efficient (while in geopandas it is a `take` on a int array)

Reply:
I think we shouldn't care about performance for the default implementation. It'll be too dependent on which optimizations the subclass can implement.

I'll go with a simple implementation like your pad_new.

Reply:
And also, if you care about performance, you probably don't have an object array under the hood but a typed array, and then this will be faster.